File: 2016/Day1/main.dart
import 'dart:collection';
import 'dart:io';

enum Direction { North, East, South, West }

class Position {
  int x;
  int y;

  Position(this.x, this.y);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Position &&
          runtimeType == other.runtimeType &&
          x == other.x &&
          y == other.y;

  @override
  int get hashCode => x.hashCode ^ y.hashCode;
}

Future<List<String>> readFile(String filename) async {
  final file = File(filename);
  final line = await file.readAsString();
  final instructions = line
      .split(',')
      .map((instr) => instr.trim())
      .where((instr) => instr.isNotEmpty)
      .toList();
  return instructions;
}

void main() async {
  final instructions = await readFile('input.txt');

  var pos = Position(0, 0);
  var dir = Direction.North;
  var visited = HashSet<Position>();
  visited.add(pos);
  var found = false;
  var firstRevisitedDistance = 0;

  for (var instr in instructions) {
    var turn = instr[0];
    var steps = int.parse(instr.substring(1));

    if (turn == 'L') {
      dir = Direction.values[(dir.index + 3) % 4]; // Left
    } else if (turn == 'R') {
      dir = Direction.values[(dir.index + 1) % 4]; // Right
    }

    for (var i = 0; i < steps; ++i) {
      switch (dir) {
        case Direction.North:
          pos = Position(pos.x, pos.y + 1);
          break;
        case Direction.East:
          pos = Position(pos.x + 1, pos.y);
          break;
        case Direction.South:
          pos = Position(pos.x, pos.y - 1);
          break;
        case Direction.West:
          pos = Position(pos.x - 1, pos.y);
          break;
      }

      if (visited.contains(pos) && !found) {
        firstRevisitedDistance = pos.x.abs() + pos.y.abs();
        found = true;
      }
      visited.add(pos);
    }
  }

  print('Part 1: ${pos.x.abs() + pos.y.abs()}');
  if (found)
    print('Part 2: $firstRevisitedDistance');
  else
    print('Part 2: Not found');
}

File: 2016/Day1/main.py
from dataclasses import dataclass
from enum import Enum
import sys


class Direction(Enum):
    NORTH = 0
    EAST = 1
    SOUTH = 2
    WEST = 3

@dataclass(frozen=True)
class Position:
    x: int
    y: int

if __name__ == "__main__":
    line = sys.stdin.readline().strip()
    instructions = [instr.strip() for instr in line.split(',')]

    pos = Position(0, 0)
    current_dir_val = Direction.NORTH.value
    visited = {pos}
    found_revisited = False
    first_revisited_distance = 0

    for instr in instructions:
        turn = instr[0]
        steps = int(instr[1:])

        if turn == 'L':
            current_dir_val = (current_dir_val + 3) % 4
        elif turn == 'R':
            current_dir_val = (current_dir_val + 1) % 4

        current_dir = Direction(current_dir_val)

        for _ in range(steps):
            x, y = pos.x, pos.y
            if current_dir == Direction.NORTH:
                y += 1
            elif current_dir == Direction.EAST:
                x += 1
            elif current_dir == Direction.SOUTH:
                y -= 1
            elif current_dir == Direction.WEST:
                x -= 1

            pos = Position(x, y)

            if pos in visited and not found_revisited:
                first_revisited_distance = abs(pos.x) + abs(pos.y)
                found_revisited = True
            visited.add(pos)

    final_distance = abs(pos.x) + abs(pos.y)
    print(f'Part 1: {final_distance}')

    if found_revisited:
        print(f'Part 2: {first_revisited_distance}')
